### PR TITLE
Add verbose mode to boilerplate

### PR DESCRIPTION
**Pull request by bowei**
hack/verify-boilerplate.sh -v will now print out why the file does
not match along with a diff if possible.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34760)

### CODE REVIEW

File: hack/boilerplate/boilerplate_test.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
+
+# Copyright 2015 The Kubernetes Authors.
```
Comment (spxtr) [inline hack/boilerplate/boilerplate_test.py]: It's 2016 :)

File: hack/boilerplate/boilerplate.py [hunk 2]
```diff
@@ -52,7 +65,8 @@ def get_refs():
 def file_passes(filename, refs, regexs):
     try:
         f = open(filename, 'r')
-    except:
+    except Exception as exc:
```
Comment (spxtr) [inline hack/boilerplate/boilerplate.py]: This should probably be more specific than `Exception`.

Comment (bowei) [inline hack/boilerplate/boilerplate.py]: I'm trying to preserve the existing behavior...

**Comment by spxtr** [comment]
What's this? A test!? Amazing!

I'm a huge fan, but I don't think we will actually run this test in CI as things stand.

File: hack/boilerplate/boilerplate_test.py (hunk 1 above)
Comment (bowei) [inline hack/boilerplate/boilerplate_test.py]: fixed

File: hack/boilerplate/boilerplate.py (hunk 2 above)
Comment (spxtr) [inline hack/boilerplate/boilerplate.py]: OK

**Comment by spxtr** [comment]
I suppose we can add a verify rule that runs these tests.

**Comment by bowei** [comment]
@spxtr : good to go or waiting for a verify script hook?

**Comment by spxtr** [comment]
I'm wary of checking in a test that doesn't run in CI. To be honest, it might be a good idea to run it in `hack/verify-boilerplate.py`.

**Comment by bowei** [comment]
I added the boilerplate unit test to the verify script

File: hack/verify-boilerplate.sh [hunk 3]
```diff
@@ -19,10 +19,21 @@ set -o nounset
...
-boiler="${KUBE_ROOT}/hack/boilerplate/boilerplate.py"
+
+boilerDir=${KUBE_ROOT}/hack/boilerplate
+boiler="${boilerDir}/boilerplate.py"
 
 files_need_boilerplate=($(${boiler} "$@"))
 
+# Run boilerplate.py unit tests
+pushd ${boilerDir} >/dev/null
+if ! python -m unittest boilerplate_test 2>/dev/null; then
```
Comment (spxtr) [inline hack/verify-boilerplate.sh]: If this fails I want to see the output.

Comment (bowei) [inline hack/verify-boilerplate.sh]: done

File: hack/verify-boilerplate.sh [hunk 4]
```diff
@@ -19,10 +19,29 @@ set -o nounset
 set -o pipefail
 
 KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
-boiler="${KUBE_ROOT}/hack/boilerplate/boilerplate.py"
+
+boilerDir=${KUBE_ROOT}/hack/boilerplate
```
Comment (spxtr) [inline hack/verify-boilerplate.sh]: `boiler_dir="${KUBE_ROOT}/hack/boilerplate"`

File: hack/verify-boilerplate.sh [hunk 5]
```diff
@@ -19,10 +19,29 @@ set -o nounset
...
+boiler="${boilerDir}/boilerplate.py"
 
 files_need_boilerplate=($(${boiler} "$@"))
 
+# Run boilerplate.py unit tests
+unitTestOut=`mktemp`
```
Comment (spxtr) [inline hack/verify-boilerplate.sh]: `unit_test_out=$(mktemp)`

File: hack/verify-boilerplate.sh [hunk 6]
```diff
@@ -19,10 +19,29 @@ set -o nounset
...
+trap cleanup EXIT
+cleanup() {
+	rm ${unitTestOut}
```
Comment (spxtr) [inline hack/verify-boilerplate.sh]: `  rm "${unit_test_out}"`

File: hack/verify-boilerplate.sh [hunk 7]
```diff
@@ -19,10 +19,29 @@ set -o nounset
...
+}
+
+pushd ${boilerDir} >/dev/null
```
Comment (spxtr) [inline hack/verify-boilerplate.sh]: quotes

File: hack/verify-boilerplate.sh [hunk 8]
```diff
@@ -19,10 +19,29 @@ set -o nounset
...
+if ! python -m unittest boilerplate_test 2>${unitTestOut}; then
```
Comment (spxtr) [inline hack/verify-boilerplate.sh]: quotes

File: hack/verify-boilerplate.sh [hunk 9]
```diff
@@ -19,10 +19,29 @@ set -o nounset
...
+	echo "boilerplate_test.py failed"
+	echo
+	cat ${unitTestOut}
```
Comment (spxtr) [inline hack/verify-boilerplate.sh]: quotes

**Review by spxtr — COMMENTED** [review]
One more round of nits, otherwise looks good.

**Comment by bowei** [comment]
updated quoting

**Comment by spxtr** [comment]
OK this seems fine to me. It's a little gross that it gets run in verify-boilerplate but I think it's fine.

**Pull request merged**